stelaval/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,176 @@
+package assignment2;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class BinaryTree<T> {
+
+  private Node<T> root;
+  private Set<Node<T>> nodeSet;
+
+  public BinaryTree() {
+    nodeSet = new HashSet<>();
+  }
+
+  /**
+   * The nodes of the binary tree are stored in a set in order to avoid duplicate values.
+   * 
+   * @param newNode is the node to be added to the set of nodes
+   */
+  public void add(Node<T> newNode) {
+    if (root == null) {
+      root = newNode;
+    }
+
+    if (nodeSet.contains(newNode)) {
+      throw new IllegalArgumentException();
+    } else {
+      nodeSet.add(newNode);
+    }
+  }
+
+  /**
+   * Prints all the ancestors of a node
+   * 
+   * @param key is a node which ancestors are printed
+   * 
+   * @throws Exception if given key is not present in a tree
+   */
+  public void printAncestors(Node<T> key) throws Exception {
+    if (nodeSet.contains(key)) {
+      containsKey(root, key);
+    } else {
+      throw new Exception("Key is out of tree");
+    }
+  }
+
+  /**
+   * Checks if a node is present in a subtree
+   * 
+   * @param currentNode is the root of a subtree
+   * 
+   * @param key is a node which ancestors will be printed
+   */
+  private boolean containsKey(Node<T> currentNode, Node<T> key) {
+    if (currentNode == null) {
+      return false;
+    } else if (currentNode.getValue().equals(key.getValue())) {
+      return true;
+    } else if (containsKey(currentNode.getLeft(), key)
+        || containsKey(currentNode.getRight(), key)) {
+      System.out.print(currentNode.getValue() + " ");
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Finds the lowest common ancestor of two nodes in a binary tree
+   * 
+   * @param node1 is the first node
+   * @param node2 is the second node
+   * 
+   * @return node that is the lowest common ancestor of node1 and node2
+   * 
+   * @throws Exception if either one or both nodes are not in the binary tree
+   */
+  public Node<T> getLowestCA(Node<T> node1, Node<T> node2) throws Exception {
+    if (nodeSet.contains(node1) && nodeSet.contains(node2)) {
+      return getCA(root, node1, node2);
+    } else {
+      throw new Exception("Nodes are not in the tree");
+    }
+  }
+
+  private Node<T> getCA(Node<T> currentNode, Node<T> node1, Node<T> node2) {

```

Comment:
`getCA` suggests that it is any common ancestor, not necessarily lowest. This method and the `getLowestCA` above have different parameters, so it would be safe to call this one `getLowestCA` as well, which would make it clear that it is still the lowest CA that we are looking for.